### ISSUE TITLE
Content Negotiation aspnetcore 2.0 web api

### ISSUE THREAD

**Issue by richb8** [aspnet/Mvc#6831]
Hi,

I have an asp.net core 2.0 mvc api application that is setup to negotiate content (JSON or XML) based on the Accept header passed in the request. However, the response from my controllers always returns JSON when i set the Accept header to 'application/xml'...

My startup config looks like this:

` // Add MVC
            services.AddMvc(options =>
            {
                options.RespectBrowserAcceptHeader = true;
                options.OutputFormatters.Add(new XmlSerializerOutputFormatter());
            })
`

Controller response:

`return Ok(response);`

The above response object is just a simple class with various string properties.

The Microsoft.AspNetCore.Mvc.Formatters.Xml NuGet package is also installed.

Am i missing something or is this a bug??

Many thanks.

**Comment by khellang** [comment]
Did you also call `AddXmlSerializerFormatters` on the MVC builder (returned from `AddMvc`)?

**Comment by richb8** [comment]
Yep.

I noticed my response class didn't have a parameterless constructor for serialization. Now i've added one, it seems to use the Accept header now however for application/xml Postman now just returns a "Could not get any response" message even though VS2017 debugging indicates the action request was a success. Strange one this..

**Comment by khellang** [comment]
AFAIK, XML serialization (using `XmlSerializer`) is pretty brittle. You have to carefully follow a bunch of rules for it to successfully serialize POCOs. Have you tried the DataContract-based as well (`AddXmlDataContractSerializerFormatters`)? What is the logs saying?

**Comment by richb8** [comment]
I'll give it a go.

Cheers

**Comment by richb8** [comment]
Not sure why I didn't look at the console logs first...

Figured it out. It's all working smoothly now.

Basically my response class had a property of type 'object' which was used for any other unknown class I wanted to add to the response. Because of this, the xml serializers couldn't serialize the unknown type.

The solutions are to:

1) Use the XmlInclude attibute to specify types that could be contain in the property.
2) Use generics.

I've opted for the latter and made my response class generic so I can set the type of that property upfront.

Cheers for your help!
